internal/identity/providers.go

```diff
@@ -192,6 +195,58 @@ func (p *Provider) IdentityFromToken(ctx context.Context, t *oauth2.Token) (*oid
 
 // Revoke enables a user to revoke her token. If the identity provider supports revocation
 // the endpoint is available, otherwise an error is thrown.
+//
+// We will attempt to get the identity provider's possible revocation endpoint from
+// their /.well-known/openid-configuration.
+// we use the revocation_endpoint or end_session_endpoint to implement
+// the generic revocation of a user's access token.
+//
+// reference:
+// Google : https://accounts.google.com/.well-known/openid-configuration
+// Azure: https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration
 func (p *Provider) Revoke(ctx context.Context, token *oauth2.Token) error {
-	return ErrRevokeNotImplemented
+	if p.RevocationURL == "" {
+		return ErrRevokeNotImplemented
+	}
+	params := url.Values{}
+	params.Add("token", token.AccessToken)
+	err := httputil.Client(ctx, http.MethodPost, p.RevocationURL, version.UserAgent(), nil, params, nil)
+	if err != nil && err != httputil.ErrTokenRevoked {
+		return err
+	}
+
+	return nil
+}
+
+// UnmarshalJSON method is needed so that provider will be able to
+// satisfy the Unmarshaler interface
+// https://pkg.go.dev/gopkg.in/square/go-jose.v2/json#Unmarshaler
+//
+// For the the go-oidc package to be able to apply the Claims method which tries to unmarshal
+// the response body from the provider's /.well-known/openid-configuration
+// on the supplied interface, and in our case, it will be the Provider.
+// An extra step is made to create this implementation because the endpoint
+// to revoke a user's access token can be revocation_endpoint or end_session_endpoint.
+// We want to be able to have only one field and not repeat any of the two possible
+// endpoints on the Provider field.
+func (p *Provider) UnmarshalJSON(b []byte) error {
+	type Alias Provider
+	t := &struct {
+		*Alias
+		EndSessionEndpoint string `json:"end_session_endpoint,omitempty"`
```

Comment:
I took another look at this. It looks like while "revocation_endpoint" is the standard rfc7009 endpoint,
"end_session_endpoint" is something entirely different and unique to azure (... in very ... azure fashion).

This endpoint is not another alias for remotely revoking a azure session (... has it always been this way? I feel like it used to be a revocation endpoint?) but instead is simply a url which can be passed the enduser and redirect them to in order to remove their session and log out. As far as I can find, there is no way to actually revoke a user's session remotely for azure.

Reply:
See https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-a-sign-out-request

Reply:
@desimone We just have to implement the revocation from https://openid.net/specs/openid-connect-frontchannel-1_0.html, we talked about on slack.